api-authentication/server/server.ts

```
@@ -0,0 +1,205 @@
+app.use(express.json());
+
+app.post('/api/auth/sign-up', async (req, res, next) => {
+  try {
+    const { username, password } = req.body as Partial<Auth>;
+    if (!username || !password) {
+      throw new ClientError(400, 'username and password are required fields');
+    }
+    const hashedPassword = await argon2.hash(password);
+    const sql = `
+      insert into "users" ("username", "hashedPassword")
+      values ($1, $2)
+      returning *
+    `;
+    const params = [username, hashedPassword];
+    const result = await db.query<User>(sql, params);
+    const [user] = result.rows;
+    res.status(201).json(user);
+  } catch (err) {
+    next(err);
+  }
+});
+
+app.post('/api/auth/sign-in', async (req, res, next) => {
+  try {
+    const { username, password } = req.body as Partial<Auth>;
+    if (!username || !password) {
+      throw new ClientError(401, 'invalid login');
+    }
+
+    /* your code starts here */
+
+    /* Query the database to find the "userId" and "hashedPassword" for the "username".
+     * If no user is found,
+     *   throw a 401: 'invalid login' error.
+     * If a user is found,
+     *   confirm that the password included in the request body matches the "hashedPassword" with `argon2.verify()`
+     *   If the password does not match,
+     *     throw a 401: 'invalid login' error.
+     *   If the password does match,
+     *     Create a payload object containing the user's "userId" and "username".
+     *     Create a new signed token with `jwt.sign()`, using the payload and your TOKEN_SECRET
+     *     (see `hashKey` above).
+     *     Send the client a 200 response containing the payload and the token.
+     */
+
+    const sql = `
+    select "userId", "hashedPassword"
+    from "users"
+    where "username" = $1
+    `;
+    const params = [username];
+    const result = await db.query<User>(sql, params);
+    const [user] = result.rows;
+    if (!user) {
+      return res.status(401).json({ Error: 'Invalid login' });
```

Since we're using error middleware, you could also `throw new ClientError(401, 'Invalid login');`.